File: src/test/java/unittests/KeywordTest.java
package unittests;

import core.scm.SCMKeyword;
import org.junit.Test;

import static java.lang.Boolean.FALSE;
import static java.lang.Boolean.TRUE;
import static org.junit.Assert.assertEquals;

public class KeywordTest extends AbstractTest {

  @Test
  public void testEvalIsKeyword() {
    assertEquals(TRUE,  eval("(keyword? :test)", env));
    assertEquals(TRUE,  eval("(keyword? :a)", env));
    assertEquals(TRUE,  eval("(keyword? :_)", env));
    assertEquals(TRUE,  eval("(keyword? (keyword \"test\")", env));
    assertEquals(TRUE,  eval("(keyword? (keyword \"a\"))", env));
    assertEquals(TRUE,  eval("(keyword? (keyword \"_\"))", env));
    assertEquals(FALSE, eval("(keyword? \"test\")", env));
    assertEquals(FALSE, eval("(keyword? #\\a)", env));
    assertEquals(FALSE, eval("(keyword? '())", env));
    assertEquals(FALSE, eval("(keyword? [])", env));
    assertEquals(FALSE, eval("(keyword? {})", env));
  }

  @Test
  public void testEvalKeywords() {
    assertEquals(SCMKeyword.intern("a"), eval(":a", env));
    assertEquals(SCMKeyword.intern("test"), eval(":test", env));
    assertEquals(SCMKeyword.intern("_"), eval(":_", env));
  }

  @Test
  public void testEvalKeywordsAsFunctions() {
    assertEquals(1L,   eval("(:a {:a 1})", env));
    assertEquals(null, eval("(:c {:a 1})", env));
    assertEquals(2L,   eval("(:c {:a 1} 2)", env));
    assertEquals(9L,   eval("(:c {:a 1, :c 9} 123)", env));
  }
}

File: src/test/java/unittests/KeywordTest.kt
package unittests

import core.scm.Keyword
import org.junit.Assert.assertEquals
import org.junit.Test

class KeywordTest : AbstractTest() {

    @Test
    fun testEvalIsKeyword() {
        assertEquals(true,  eval("(keyword? :test)"))
        assertEquals(true,  eval("(keyword? :a)"))
        assertEquals(true,  eval("(keyword? :_)"))
        assertEquals(true,  eval("(keyword? (keyword \"test\")"))
        assertEquals(true,  eval("(keyword? (keyword \"a\"))"))
        assertEquals(true,  eval("(keyword? (keyword \"_\"))"))
        assertEquals(false, eval("(keyword? \"test\")"))
        assertEquals(false, eval("(keyword? #\\a)"))
        assertEquals(false, eval("(keyword? '())"))
        assertEquals(false, eval("(keyword? [])"))
        assertEquals(false, eval("(keyword? {})"))
    }

    @Test
    fun testEvalKeywords() {
        assertEquals(Keyword.intern("a"), eval(":a"))
        assertEquals(Keyword.intern("test"), eval(":test"))
        assertEquals(Keyword.intern("_"), eval(":_"))
    }

    @Test
    fun testEvalKeywordsAsFunctions() {
        assertEquals(1L,   eval("(:a {:a 1})"))
        assertEquals(null, eval("(:c {:a 1})"))
        assertEquals(2L,   eval("(:c {:a 1} 2)"))
        assertEquals(9L,   eval("(:c {:a 1, :c 9} 123)"))
    }
}
